File: python/graphs/bfs_production.py
from collections import deque
from collections.abc import Iterable


Cell = tuple[int, int]


def bfs_path(
    width: int,
    height: int,
    walls: Iterable[Cell],
    start: Cell,
    target: Cell,
) -> list[Cell]:
    if width <= 0 or height <= 0:
        raise ValueError("grid dimensions must be positive")

    def inside(cell: Cell) -> bool:
        return 0 <= cell[0] < height and 0 <= cell[1] < width

    if not inside(start) or not inside(target):
        raise ValueError("start and target must be inside the grid")

    total = width * height
    blocked = bytearray(total)
    for wall in walls:
        if inside(wall):
            blocked[wall[0] * width + wall[1]] = 1

    start_id = start[0] * width + start[1]
    target_id = target[0] * width + target[1]
    if blocked[start_id] or blocked[target_id]:
        return []

    parent = [-1] * total
    visited = bytearray(total)
    queue = deque([start_id])
    visited[start_id] = 1

    while queue:
        current_id = queue.popleft()
        if current_id == target_id:
            path = []
            while current_id != -1:
                path.append(divmod(current_id, width))
                current_id = parent[current_id]
            path.reverse()
            return path

        row, col = divmod(current_id, width)
        for next_row, next_col in (
            (row - 1, col),
            (row, col + 1),
            (row + 1, col),
            (row, col - 1),
        ):
            if not (0 <= next_row < height and 0 <= next_col < width):
                continue

            next_id = next_row * width + next_col
            if blocked[next_id] or visited[next_id]:
                continue

            visited[next_id] = 1
            parent[next_id] = current_id
            queue.append(next_id)

    return []

File: java/graphs/BfsProduction.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Deque;
import java.util.List;

public class BfsProduction {
    record Cell(int row, int col) {}

    public static List<Cell> bfsPath(boolean[][] walls, Cell start, Cell target) {
        int height = walls.length;
        int width = walls[0].length;
        validateCell(start, width, height);
        validateCell(target, width, height);

        if (walls[start.row()][start.col()] || walls[target.row()][target.col()]) {
            return List.of();
        }

        int total = width * height;
        int startId = start.row() * width + start.col();
        int targetId = target.row() * width + target.col();
        boolean[] visited = new boolean[total];
        int[] parent = new int[total];
        Deque<Integer> queue = new ArrayDeque<>();

        java.util.Arrays.fill(parent, -1);
        queue.addLast(startId);
        visited[startId] = true;

        while (!queue.isEmpty()) {
            int currentId = queue.removeFirst();
            if (currentId == targetId) {
                List<Cell> path = new ArrayList<>();
                for (int stepId = targetId; stepId != -1; stepId = parent[stepId]) {
                    path.add(new Cell(stepId / width, stepId % width));
                }
                Collections.reverse(path);
                return path;
            }

            int row = currentId / width;
            int col = currentId % width;
            int[][] moves = {
                {-1, 0},
                {0, 1},
                {1, 0},
                {0, -1}
            };

            for (int[] move : moves) {
                int nextRow = row + move[0];
                int nextCol = col + move[1];
                if (nextRow < 0 || nextRow >= height || nextCol < 0 || nextCol >= width) {
                    continue;
                }

                int nextId = nextRow * width + nextCol;
                if (walls[nextRow][nextCol] || visited[nextId]) {
                    continue;
                }

                visited[nextId] = true;
                parent[nextId] = currentId;
                queue.addLast(nextId);
            }
        }

        return List.of();
    }

    private static void validateCell(Cell cell, int width, int height) {
        if (cell.row() < 0 || cell.row() >= height || cell.col() < 0 || cell.col() >= width) {
            throw new IllegalArgumentException("start and target must be inside the grid");
        }
    }
}
